File: frontend/src/api/agent.ts
/** Vision, the in-app agent: streamed turns and decisions (NDJSON). */

import { ApiError, http, normalise } from './http'

export interface PageContext {
  pathname: string
  title?: string
  area?: string
  scope?: Record<string, unknown> | null
  visible_text: string
}

export interface AgentProposal {
  id: string
  tool: string
  label: string
  summary: string
  detail: string
  effects: string[]
  arguments: Record<string, unknown>
  payloadHash: string
  spends: Record<string, number>
  irreversible: boolean
  status: string
}

export type AgentEvent =
  | { type: 'start'; threadId: number }
  | { type: 'text'; delta: string }
  | { type: 'thinking'; delta: string }
  | { type: 'status'; delta: string }
  | {
      type: 'tool_start'
      id: string
      tool: string
      label: string
      args: Record<string, unknown>
    }
  | {
      type: 'tool_end'
      id: string
      status: string
      httpStatus: number | null
      preview: string
    }
  | { type: 'proposal'; proposal: AgentProposal }
  | { type: 'navigate'; to: string }
  | { type: 'error'; message: string }
  | { type: 'done'; threadId: number }

async function stream(
  path: string,
  body: unknown,
  onEvent: (e: AgentEvent) => void,
  signal?: AbortSignal,
) {
  const response = await fetch(path, {
    method: 'POST',
    headers: { 'X-Harness-Client': '1', 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
    signal: signal ?? null,
  })
  if (!response.ok || !response.body) {
    const raw = await response.text()
    let parsed: unknown = raw
    try {
      parsed = JSON.parse(raw)
    } catch {
      // plain text
    }
    throw new ApiError(response.status, normalise(response.status, parsed))
  }
  const reader = response.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''
  for (;;) {
    const { value, done } = await reader.read()
    if (done) break
    buffer += decoder.decode(value, { stream: true })
    const lines = buffer.split('\n')
    buffer = lines.pop() ?? ''
    for (const line of lines) if (line.trim()) onEvent(JSON.parse(line) as AgentEvent)
  }
  if (buffer.trim()) onEvent(JSON.parse(buffer) as AgentEvent)
}

export type PermissionMode = 'ask' | 'auto'

export interface Permissions {
  mode: PermissionMode
  alwaysYours: { action: string; summary: string; tier: string }[]
}

export interface GoalVerdict {
  turn: number
  verdict: string
  reason: string
  at: number
}

export interface Goal {
  objective: string
  doneWhen: string
  status: string
  running: boolean
  stoppedReason: string
  createdAt: number
  turns: number
  maxTurns: number
  lastVerdict: string
  lastReason: string
  waitUntil: number | null
  threadId: number | null
  history: GoalVerdict[]
  budgets: Record<string, number>
  spent: Record<string, number>
  remaining: Record<string, number | null>
}

export type GoalStep =
  | { action: 'none'; goal: Goal | null }
  | { action: 'continue'; prompt: string; goal: Goal }
  | { action: 'wait'; seconds: number; prompt: string; goal: Goal }
  | { action: 'await_approval'; goal: Goal }
  | { action: 'stop'; goal: Goal }

export const agent = {
  goal: () => http.get<{ goal: Goal | null }>('/api/agent/goal'),
  setGoal: (body: {
    objective: string
    done_when: string
    brain_simulations: number
    max_turns: number
  }) => http.put<{ goal: Goal }>('/api/agent/goal', body),
  goalStep: (threadId: number | null) =>
    http.post<GoalStep>('/api/agent/goal/step', { thread_id: threadId }),
  pauseGoal: () => http.post<{ goal: Goal | null }>('/api/agent/goal/pause'),
  resumeGoal: () => http.post<{ goal: Goal | null }>('/api/agent/goal/resume'),
  clearGoal: () => http.del<{ goal: null }>('/api/agent/goal'),
  permissions: () => http.get<Permissions>('/api/agent/permissions'),
  setPermissions: (mode: PermissionMode) =>
    http.put<Permissions>('/api/agent/permissions', { mode }),
  turn: (
    body: { text: string; thread_id: number | null; context: PageContext },
    onEvent: (e: AgentEvent) => void,
    signal?: AbortSignal,
  ) => stream('/api/agent/turn', body, onEvent, signal),
  decide: (
    id: string,
    body: { approve: boolean; payload_hash: string; context: PageContext },
    onEvent: (e: AgentEvent) => void,
    signal?: AbortSignal,
  ) => stream(`/api/agent/proposals/${encodeURIComponent(id)}/decide`, body, onEvent, signal),
}

File: backend/src/alpha_harness/api/agent.py
"""Vision, the in-app agent: streamed turns and decisions (NDJSON), plus the action catalog."""

from __future__ import annotations

import json
from collections.abc import AsyncIterator
from typing import Annotated, Any, Literal

from fastapi import APIRouter, Query, Request
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from ..agent.goals import Goal
from ..agent.loop import AgentService

router = APIRouter(prefix="/api/agent", tags=["agent"])


class PageContext(BaseModel):
    pathname: str = ""
    title: str | None = None
    area: str | None = None
    scope: dict[str, Any] | None = None
    visible_text: str = Field(default="", max_length=20_000)


class TurnRequest(BaseModel):
    text: str = Field(min_length=1, max_length=8_000)
    thread_id: int | None = None
    context: PageContext = Field(default_factory=PageContext)


class DecisionRequest(BaseModel):
    approve: bool
    payload_hash: str | None = None
    context: PageContext = Field(default_factory=PageContext)


async def _service(request: Request) -> AgentService:
    service = getattr(request.app.state, "agent", None)
    if service is None:
        service = AgentService(request.app, request.app.state.harness)
        request.app.state.agent = service
    await service.load()
    return service


def _context(ctx: PageContext) -> dict[str, Any]:
    return {
        "pathname": ctx.pathname,
        "title": ctx.title,
        "area": ctx.area,
        "scope": ctx.scope,
        "visibleText": ctx.visible_text,
    }


def _ndjson(events: AsyncIterator[dict[str, Any]]) -> StreamingResponse:
    async def body() -> AsyncIterator[bytes]:
        async for event in events:
            yield (json.dumps(event, default=str) + "\n").encode()

    return StreamingResponse(
        body(),
        media_type="application/x-ndjson",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )


@router.post("/turn")
async def turn(payload: TurnRequest, request: Request) -> StreamingResponse:
    service = await _service(request)
    return _ndjson(service.turn(payload.thread_id, payload.text, _context(payload.context)))


@router.post("/proposals/{proposal_id}/decide")
async def decide(proposal_id: str, payload: DecisionRequest, request: Request) -> StreamingResponse:
    service = await _service(request)
    return _ndjson(
        service.decide(
            proposal_id, payload.payload_hash, payload.approve, _context(payload.context)
        )
    )


class GoalRequest(BaseModel):
    """What to pursue, and the most it may cost. 0 on a budget means no ceiling."""

    objective: str = Field(min_length=3, max_length=600)
    done_when: str = Field(default="", max_length=600, description="How to tell it is met")
    max_turns: int = Field(default=20, ge=1, le=100)
    brain_simulations: int = Field(default=0, ge=0, le=5_000)
    llm_requests: int = Field(default=0, ge=0, le=5_000)
    correlation_jobs: int = Field(default=0, ge=0, le=500)
    #: The conversation to carry on in, and the page the person is on.
    thread_id: int | None = None
    context: PageContext = Field(default_factory=PageContext)


@router.get("/goal")
async def get_goal(request: Request) -> dict[str, Any]:
    goal = (await _service(request)).goals.goal
    return {"goal": goal.to_dict() if goal else None}


def _publish(service: AgentService, kind: str, text: str) -> dict[str, Any]:
    goal = service.goals.goal
    view = goal.to_dict() if goal else None
    service.bus.publish({"type": "goal", "goal": view})
    service.bus.publish({"type": "loop", "kind": kind, "text": text})
    return {"goal": view}


@router.put("/goal")
async def set_goal(payload: GoalRequest, request: Request) -> dict[str, Any]:
    """Set the standing objective and start working on it, in the backend: closing the tab
    does not stop it. The person's to set; Vision has no action for it."""
    service = await _service(request)
    await service.runner.stop()
    goal = service.goals.set(
        Goal(
            objective=payload.objective.strip(),
            done_when=payload.done_when.strip(),
            max_turns=payload.max_turns,
            brain_simulations=payload.brain_simulations,
            llm_requests=payload.llm_requests,
            correlation_jobs=payload.correlation_jobs,
            thread_id=payload.thread_id,
            context=_context(payload.context),
        )
    )
    service.thread_for_goal()
    await service.persist()
    service.bus.mark_goal_start()
    result = _publish(service, "start", f"Goal started: {goal.objective}")
    service.runner.start(service.runner.prompt_for("start"))
    return result


@router.post("/goal/pause")
async def pause_goal(request: Request) -> dict[str, Any]:
    service = await _service(request)
    await service.runner.stop()
    goal = service.goals.goal
    if goal is not None and goal.running:
        goal.status = "paused"
        goal.note("paused", "Paused by the person.")
    await service.persist()
    return _publish(
        service, "stop", "Paused. Carry on with /goal resume, or /goal clear to drop it."
    )


@router.post("/goal/resume")
async def resume_goal(request: Request) -> dict[str, Any]:
    """Resume a paused goal, or give an ended one a fresh turn budget."""
    service = await _service(request)
    goal = service.goals.goal
    if goal is None:
        return {"goal": None}
    if not goal.running and goal.status != "met":
        if goal.turns >= goal.max_turns:
            goal.max_turns = goal.turns + 20
        goal.status, goal.stopped_reason, goal.idle_turns, goal.judge_failures = "active", "", 0, 0
        goal.note("resumed", "Resumed by the person.")
    await service.persist()
    result = _publish(service, "start", "Goal resumed.")
    service.runner.start(service.runner.prompt_for("resume"))
    return result


@router.delete("/goal")
async def clear_goal(request: Request) -> dict[str, Any]:
    service = await _service(request)
    await service.runner.stop()
    service.goals.clear("cleared by the person")
    await service.persist()
    return _publish(service, "stop", "Goal cleared.")


@router.get("/events")
async def events(
    request: Request, since: Annotated[int, Query(ge=-1)] = -1
) -> StreamingResponse:
    """Everything Vision does on its own (goal turns, verdicts, waits), as NDJSON, from
    ``since`` onward and then live. A reopened panel catches up from where it left off."""
    service = await _service(request)
    # -1: a panel opening fresh. Replay the current goal's story, not the whole log.
    start = service.bus.goal_from if since < 0 else since

    async def body() -> AsyncIterator[bytes]:
        async for event in service.bus.follow(start):
            if await request.is_disconnected():
                return
            yield (json.dumps(event, default=str) + "\n").encode()

    return StreamingResponse(
        body(),
        media_type="application/x-ndjson",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )


class PermissionsRequest(BaseModel):
    mode: Literal["ask", "auto"]


@router.get("/permissions")
async def get_permissions(request: Request) -> dict[str, Any]:
    service = await _service(request)
    return {"mode": service.permissions.mode, "alwaysYours": _always_yours(service)}


@router.put("/permissions")
async def set_permissions(payload: PermissionsRequest, request: Request) -> dict[str, Any]:
    """Set by the person in the UI. Excluded from Vision's catalog, so it cannot set its own."""
    service = await _service(request)
    service.permissions.set(payload.mode)
    return {"mode": service.permissions.mode, "alwaysYours": _always_yours(service)}


def _always_yours(service: AgentService) -> list[dict[str, str]]:
    rows = [
        {"action": f"{e['method']} {e['path']}", "summary": e["summary"], "tier": e["tier"]}
        for e in service.index.values()
        if e["tier"] in {"human_only", "blocked"}
    ]
    rows.append(
        {
            "action": "Submit an alpha to BRAIN",
            "summary": "Blocked in the BRAIN client",
            "tier": "blocked",
        }
    )
    return rows


@router.get("/proposals")
async def proposals(request: Request) -> list[dict[str, Any]]:
    return [p.to_dict() for p in (await _service(request)).gate.pending()]


@router.get("/actions")
async def catalog(request: Request) -> list[dict[str, Any]]:
    return (await _service(request)).catalog()
